core/jvm/test/PoolingTest.kt

```diff
@@ -40,4 +45,107 @@ class PoolingTest {
         otherBuffer.clear()
         assertTrue(poolSize < SegmentPool.byteCount)
     }
+
+    @Test
+    fun segmentAcquisitionAndRelease() {
+        val secondTierSize = SegmentPool.SECOND_LEVEL_POOL_TOTAL_SIZE
+        val firstTierSize = SegmentPool.MAX_SIZE
+
+        // fill the pool by requiring max possible segments count and then
+        // releasing them all
+        val segments = mutableSetOf<Segment>()
+        var size = 0
+        while (size < secondTierSize + firstTierSize) {
+            val segment = SegmentPool.take()
+            size += segment.dataAsByteArray(false).size
+            segments.add(segment)
+        }
+        segments.forEach(SegmentPool::recycle)
+
+
+        // take the same number of segments again and check that nothing new was allocated
+        val segments2 = mutableSetOf<Segment>()
+        size = 0
+        while (size < secondTierSize + firstTierSize) {
+            val segment = SegmentPool.take()
+            size += segment.remainingCapacity
+            segments2.add(segment)
+        }
+        segments2.forEach(SegmentPool::recycle)
+
+        assertEquals(segments, segments2)
+    }
+
+    @Test
+    fun contendedUseTest() {
+        val threadsCount = Runtime.getRuntime().availableProcessors()
+        val segmentsPerThread = SegmentPool.SECOND_LEVEL_POOL_TOTAL_SIZE / Segment.SIZE / threadsCount
```

Comment:
It turns out that we only take into account fully filled segments?

Reply:
Pool size reflects maximum number of segments it'll hold once they returned back into pool. When segments returned, they are all empty, so their whole size contributes into the pool size.